roofit/roofit/src/RooLagrangianMorphFunc.cxx

```diff
@@ -0,0 +1,3572 @@
+  this->collectInputs(file);
+
+  cache->buildMatrix(this->_config.getParamCards(),
+                     this->_config.getFlagValues(), this->_flags);
+  this->updateSampleWeights();
+
+  closeFile(file);
+  return true;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// setup the morphing function with a predefined inverse matrix
+/// call this function *before* any other after creating the object
+
+bool RooLagrangianMorphFunc::useCoefficients(const TMatrixD &inverse) {
+  RooLagrangianMorphFunc::CacheElem *cache =
+      (RooLagrangianMorphFunc::CacheElem *)_cacheMgr.getObj(0, (RooArgSet *)0);
+  Matrix m = makeSuperMatrix(inverse);
+  if (cache) {
+    std::string filename = this->_config.getFileName();
+    cache->_inverse = m;
+    TDirectory *file = openFile(filename.c_str());
+    if (!file) {
+      coutE(InputArguments)
+          << "unable to open file '" << filename << "'!" << std::endl;
+      return false;
+    }
+    cxcoutP(InputArguments) << "reading parameter sets." << std::endl;
+
+    this->readParameters(file);
+    checkNameConflict(this->_config.getParamCards(), this->_operators);
+    this->collectInputs(file);
+
+    // then, update the weights in the morphing function
+    this->updateSampleWeights();
+
+    closeFile(file);
+  } else {
+    cache = RooLagrangianMorphFunc::CacheElem::createCache(this, m);
+    if (!cache)
+      coutE(Caching) << "unable to create cache!" << std::endl;
+    this->_cacheMgr.setObj(0, 0, cache, 0);
+  }
+  return true;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// setup the morphing function with a predefined inverse matrix
+// call this function *before* any other after creating the object
+
+bool RooLagrangianMorphFunc::useCoefficients(const char *filename) {
+  RooLagrangianMorphFunc::CacheElem *cache =
+      (RooLagrangianMorphFunc::CacheElem *)_cacheMgr.getObj(0, (RooArgSet *)0);
+  if (cache) {
+    return false;
+  }
+  cache = RooLagrangianMorphFunc::CacheElem::createCache(
+      this, readMatrixFromFileT<Matrix>(filename));
+  if (!cache)
+    coutE(Caching) << "unable to create cache!" << std::endl;
+  this->_cacheMgr.setObj(0, 0, cache, 0);
+  return true;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// write the inverse matrix to a file
+
+bool RooLagrangianMorphFunc::writeCoefficients(const char *filename) {
+  auto cache = this->getCache(_curNormSet);
+  if (!cache)
+    return false;
+  writeMatrixToFileT(cache->_inverse, filename);
+  return true;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// retrieve the cache object
+
+typename RooLagrangianMorphFunc::CacheElem *
+RooLagrangianMorphFunc::getCache(const RooArgSet * /*nset*/) const {
+  RooLagrangianMorphFunc::CacheElem *cache =
+      (RooLagrangianMorphFunc::CacheElem *)_cacheMgr.getObj(0, (RooArgSet *)0);
```

Comment:
```suggestion
  auto cache = static_cast<RooLagrangianMorphFunc::CacheElem *>(_cacheMgr.getObj(0, (RooArgSet *)0));
```
* avoid C-style casts
* you can use `auto` if you already do a cast so you don't have to write potentially long class names redundantly